src/uct/base/uct_iface.c

```diff
@@ -553,8 +554,20 @@ ucs_status_t uct_iface_reject(uct_iface_h iface,
 
 ucs_status_t uct_ep_create(const uct_ep_params_t *params, uct_ep_h *ep_p)
 {
+    uct_base_iface_t *base_iface;
+    ucs_status_t status;
+
     if (params->field_mask & UCT_EP_PARAM_FIELD_IFACE) {
-        return params->iface->ops.ep_create(params, ep_p);
+        status = params->iface->ops.ep_create(params, ep_p);
+        if (status != UCS_OK) {
+            return status;
+        }
+
+        /* Create endpoint VFS node on demand to avoid memory bloat */
+        base_iface = ucs_derived_of(params->iface, uct_base_iface_t);
+        ucs_vfs_obj_set_dirty(params->iface,
+                              base_iface->internal_ops->iface_vfs_refresh);
```

Comment:
should do this on "ep" directory which is created in RC transport, so no need to add new callback function to internal_ops

Reply:
but this will allow us extend it for other transports too

Reply:
so can add helper function uct_iface_vfs_add_ep_dir() which accepts the refresh callback as parameter

Reply:
> so can add helper function uct_iface_vfs_add_ep_dir() which accepts the refresh callback as parameter

but then it should be called from transport iface init function?

Reply:
sorry, I meant - which code is responsible to call `uct_iface_vfs_add_ep_dir()` function?

Reply:
@yosefe if I understand correctly, we should introduce `uct_iface_vfs_add_ep_dir()` in UCT/BASE code and use in EP's init function of transports. But then it means each transport should be responsible calling this function.
Here I've implemented the same way as we do for UCT iface objects (VFS is created in common function `uct_iface_open()` instead of transport-specific code)

Reply:
the idea is nice but pls consider following points:
- Better not add extra internal_ops which every transport will have to initialize (to empty function)
- After this PR, iface global VFS files (such as CQ stuff) are created when iface is created. but it's better to add them on-demand.
- the names (init/refresh) do not indicate whet they are really doing. why need to separate "init" and "refresh"?
 better add common infra to show **endpoints** of an iface. including, for example, creating the endpoints directory, populating it on demand, and removing the endpoint VFS entry when the endpoint is destroyed.

Reply:
ok, thanks for the explanation.
So, my plan is to:
- revert code the previous state
- add `ep_vfs_refresh` in internal ops. which is set in `ucs_vfs_obj_set_dirty()` for EP object when creating new endpoint
is my understanding correct?

Reply:
yes, and i think we need to create "iface/%p/eps" directory in base , after it calls iface_vfs_refresh(). and set it's refresh callback to internal_ops->ep_vfs_refresh

Reply:
why need "iface/%/eps" directory? I guess we should have "iface/%p/ep/%p" as it is done in UCP

Reply:
> after it calls iface_vfs_refresh(). and set its refresh callback to internal_ops->ep_vfs_refresh

not sure that I understand - should we set `internal->ep_vfs_refresh` for each "ep/%p" directory inside `iface_vfs_refresh` callback by going over all EPs?

Reply:
sorry, i meant `iface/%p/ep`
the refresh cb should be set for the "ep" directory. it will go over all iface's eps and create "ep/%p".

Reply:
We pass an `obj` associated with the `node` to `refresh_cb`.
E.g. we mark `/ucp/worker/worker_name` as dirty in method that creates UCP ep. Then when we `touch` the directory, refresh callback is executed. And we pass worker object represented by the directory. Using the object we get actual list of UCP endpoints and add required directories and files.

I think here we should use the same approach. We should mark dirty a directory which represents a parent object (parent means that the object can be used to get list of child objects). Thus we should mark `uct/iface/iface_ptr` folder dirty.

Reply:
here we have 2 options:
1. single level of refresh: mark "iface/%p" as dirty. so all eps should be added from iface_vfs_refresh(). and no need to add new internal_ops
2. two levels of refresh: mark "iface/%p/eps" as dirty. all eps should be added from a separate refresh callback. 

approach (1) is simpler, and approach (2) allows accessing iface fields w/o enumerating all its endpoints

Reply:
The second option requires additional features in VFS infra.
`ucs_vfs_obj_set_dirty` has two arguments: object and refresh callback. Calling the method we mark dirty a directory representing the object in VFS structure. Thus in current implementation we cannot mark dirty `iface/%p/eps`.

Reply:
yes, you are right. we cannot set directly on a directory without an object. so need to return to option (1)

Reply:
why we can’t mark “ep/%p” as dirty and set internal_ops->ep_vfs_refresh callback for it?
It will be the same as we have for internal_ops->iface_vfs_refresh which is set for “iface/%p”

Reply:
the idea with mark dirty / refresh is not create many objects (per every ep) 
and if we create "ep/%p" anyway for each ep there is no point to use mark dirty scheme

Reply:
ok, just to summarize:
1. leave the code as is
2. go over all eps in `iface_vfs_refresh()` callback: create "ep/%p" directory and add files (qp_num, ci, pi and etc)
is it correct?

Reply:
+1

Reply:
done